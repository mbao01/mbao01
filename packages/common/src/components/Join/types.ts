import { type VariantProps } from "../../libs";
import { getJoinClasses, getJoinItemClasses } from "./constants";

type FormElements = keyof React.JSX.IntrinsicElements;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type As<P = any> =
  | FormElements
  | React.ComponentType<P>
  | React.ForwardRefExoticComponent<P>;

type AllowedElements = Pick<React.JSX.IntrinsicElements, FormElements>;

type AsProps<T extends As> = T extends FormElements
  ? AllowedElements[T]
  : T extends React.ForwardRefExoticComponent<infer P>
    ? P
    : T extends React.ComponentType<infer P>
      ? P
      : never;

export type JoinProps<T extends As> = AsProps<T> &
  VariantProps<typeof getJoinClasses> & {
    as: T;
  };

export type JoinItemProps<T extends As> = AsProps<T> &
  VariantProps<typeof getJoinItemClasses> & {
    as: T;
  };
