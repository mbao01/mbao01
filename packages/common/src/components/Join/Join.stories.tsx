import type { Meta, StoryObj } from "@storybook/react-vite";
import { Badge } from "../Badge";
import { Button } from "../Button";
import { Indicator } from "../Indicator";
import { Join } from "./Join";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Join,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Join>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    as: "span",
  },
  render: (args) => {
    return (
      <Join {...args}>
        <Join.Item {...args}>ðŸ˜ƒ</Join.Item>
        <Join.Item {...args}>ðŸ˜…</Join.Item>
      </Join>
    );
  },
};

export const JoinTwoButtons: Story = {
  args: {
    as: "span",
  },
  render: (args) => {
    return (
      <Join {...args}>
        <Join.Item as={Button} size="lg" variant="success">
          Hello
        </Join.Item>
        <Join.Item as={Button} size="lg" outline variant="primary">
          World
        </Join.Item>
      </Join>
    );
  },
};

export const JoinTwoButtonWiths: Story = {
  args: {
    as: "span",
  },
  render: (args) => {
    return (
      <Join {...args}>
        <Join.Item as={Button} size="lg" outline variant="success">
          Hello
        </Join.Item>
        <Indicator as="span" position={["start", "top"]}>
          <Indicator.Item as={Badge} size="sm" variant="error">
            2
          </Indicator.Item>
          <Join.Item as={Button} size="lg" variant="primary">
            World
          </Join.Item>
        </Indicator>
      </Join>
    );
  },
};
