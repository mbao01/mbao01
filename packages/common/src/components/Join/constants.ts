import { cva } from "../../libs";

export const getJoinClasses = cva("join");

export const getJoinItemClasses = cva("rounded-[auto] join-item");
