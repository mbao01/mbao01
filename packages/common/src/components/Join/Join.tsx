import { Slot } from "@radix-ui/react-slot";
import type { As, JoinItemProps, JoinProps } from "./types";
import { cn } from "../../utilities";
import { getJoinClasses, getJoinItemClasses } from "./constants";

const Join = <T extends As>({ as, className, children, ...props }: JoinProps<T>) => {
  const SlotChild = as;

  return (
    <Slot className={cn(getJoinClasses(), className)} {...props}>
      <SlotChild>{children}</SlotChild>
    </Slot>
  );
};

const JoinItem = <T extends As>({ as, className, children, ...props }: JoinItemProps<T>) => {
  const SlotChild = as;

  return (
    <Slot className={cn(getJoinItemClasses(), className)} {...props}>
      <SlotChild>{children}</SlotChild>
    </Slot>
  );
};

Join.Item = JoinItem;

export { Join };
