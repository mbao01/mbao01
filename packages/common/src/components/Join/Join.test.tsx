import { render, screen } from "@testing-library/react";
import { Join } from "./Join";

describe("Join", () => {
  it("renders a basic join", () => {
    const { asFragment } = render(
      <Join as="span" data-testid="content">
        <Join.Item as="span">First</Join.Item>
        <Join.Item as="span">Second</Join.Item>
      </Join>
    );

    expect(screen.getByTestId("content")).toHaveTextContent("FirstSecond");
    expect(asFragment()).toMatchSnapshot();
  });
});
