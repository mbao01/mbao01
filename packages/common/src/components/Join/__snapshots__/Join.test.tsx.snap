// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Join > renders a basic join 1`] = `
<DocumentFragment>
  <span
    class="join"
    data-testid="content"
  >
    <span
      class="rounded-[auto] join-item"
    >
      First
    </span>
    <span
      class="rounded-[auto] join-item"
    >
      Second
    </span>
  </span>
</DocumentFragment>
`;
